### ISSUE TITLE
How to limit the number of options for setting parameters to avoid Ddos-attacks?

### ISSUE THREAD

**Issue by monochromer** [quix-labs/caddy-image-processor#9]
For example, attacker can use something like that:
```
https://example.com/image.jpg?w=401
https://example.com/image.jpg?w=402
https://example.com/image.jpg?w=403
...
https://example.com/image.jpg?w=9999
```

**Comment by alancolant** [comment]
I don't fully understand the request. Is your need to apply throttling on the number of incoming requests, for example, limiting each IP to 30 requests per minute?

In that case, there are plugins like [caddy-ratelimit](https://github.com/mholt/caddy-ratelimit) that would do the job.

Or are you imagining a limit (like a range) for certain parameters?

For example, specifying that the width can only be between 100px and 1000px.  
The same could apply for other parameters.

This would help limit CPU time, as a request for 10_000px could easily consume a lot of CPU time and overload memory as well.

I also imagine putting a time limit on generation, to stop the process if it takes more than 5 seconds, for example. I need to look into this in more depth to see if it's possible, though I'm not sure.

If you have other ideas, I'm listening!

**Comment by monochromer** [comment]
> specifying that the width can only be between 100px and 1000px.

Yes, I meant it. But also with discrete step. For example, only this values may be applied - `100`, `200`, ..., `1000`.

**Comment by alancolant** [comment]
> > specifying that the width can only be between 100px and 1000px.
> 
> Yes, I meant it. But also with discrete step. For example, only this values may be applied - `100`, `200`, ..., `1000`.

Okay, I will think about a solution for this.

It will take some time, and I think I will try to add several layers of control.

For now, I will work on implementing these two restrictions:  
* Disabling soecific parameters.  
* Limiting width and height within certain bounds. I will consider discrete values later.  

We need to decide whether we prefer to return a 400 error if one of these conditions is not met, simply ignore it, or serve the original image by bypassing the processing.  

I have other ideas for improving security, but they will take much longer to implement.

**Comment by alancolant** [comment]
@monochromer 

Implemented, see README and use 2.3.1 release, here is an example:
```

localhost:80 {
	import common
	root test-dataset
	file_server
	image_processor {
	    on_fail bypass
            security {
                on_security_fail bypass
                constraints {
                    h range 60 480
                    w {
                        values 60 130 240 480 637
                    }
                }
           }
       }
}
```

**Comment by monochromer** [comment]
Thanks!
